src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/Reliability/RetryPolicy.cs

```diff
@@ -0,0 +1,241 @@
+        /// </summary>
+        public event EventHandler<RetryingEventArgs> Retrying;
+
+        /// <summary>
+        /// Gets the retry strategy.
+        /// </summary>
+        public RetryStrategy RetryStrategy { get; private set; }
+
+        /// <summary>
+        /// Gets the instance of the error detection strategy.
+        /// </summary>
+        public ITransientErrorDetectionStrategy ErrorDetectionStrategy { get; private set; }
+
+        /// <summary>
+        /// Repetitively executes the specified action while it satisfies the current retry policy.
+        /// </summary>
+        /// <param name="action">A delegate that represents the executable action that doesn't return any results.</param>
+        public virtual void ExecuteAction(Action action)
+        {
+            this.ExecuteAction(() => { action(); return default(object); });
+        }
+
+        /// <summary>
+        /// Repetitively executes the specified action while it satisfies the current retry policy.
+        /// </summary>
+        /// <typeparam name="TResult">The type of result expected from the executable action.</typeparam>
+        /// <param name="func">A delegate that represents the executable action that returns the result of type <typeparamref name="TResult"/>.</param>
+        /// <returns>The result from the action.</returns>
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1062:Validate arguments of public methods", MessageId = "0", Justification = "Validated with Guard")]
+        public virtual TResult ExecuteAction<TResult>(Func<TResult> func)
+        {
+            int retryCount = 0;
+            TimeSpan delay = TimeSpan.Zero;
+            Exception lastError;
+
+            var shouldRetry = this.RetryStrategy.GetShouldRetry();
+
+            for (;;)
+            {
+                lastError = null;
+
+                try
+                {
+                    return func();
+                }
+                catch (Exception ex)
+                {
+                    lastError = ex;
+
+                    if (!(this.ErrorDetectionStrategy.IsTransient(lastError) && shouldRetry(retryCount++, lastError, out delay)))
+                    {
+                        throw;
+                    }
+                }
+
+                // Perform an extra check in the delay interval. Should prevent from accidentally ending up with the value of -1 that will block a thread indefinitely. 
+                // In addition, any other negative numbers will cause an ArgumentOutOfRangeException fault that will be thrown by Thread.Sleep.
+                if (delay.TotalMilliseconds < 0)
+                {
+                    delay = TimeSpan.Zero;
+                }
+
+                this.OnRetrying(retryCount, lastError, delay);
+
+                if (retryCount > 0)
+                {
+                    Task.Delay(delay).Wait();
```

Comment:
Wouldn't this potentially need a new thread for execution causing the need for more threads from the pool and causing application scalability issues?

Reply:
Would you recommend a Thread.Sleep instead? How would it be performance-wise? Technically speaking, these events should be relatively rare, so i'm not sure how this can make a difference.

Reply:
Task.Delay() will spin up a task which may potentially run on a different thread and utilize another thread from the threadpool. And Wait() would simply wait for the delay to complete. If this method is only called synchronously, then Thread.Sleep is the way to go, which would call the same thread to be delayed instead of spinning up another one. 

If there is something going wrong causing a lot of retries to happen, then I can imagine a bunch of threads being utilized from the thread pool, to just wait on nothing, causing further delays in queuing up .Net tasks. This could potentially impact scalability of the feature.

Reply:
Makes sense, thanks!

Reply:
While reading this, it occured to me that integrating a circuit breaker pattern could be a follow-up-ask for this feature.
I.e. often these exceptions will be encountered by multiple threads running, all trying to e.g. reconnect to an (Azure) SQL. Having each and every thread going through the same retry logic might even (in case of resource starvation) aggravate the problem. Having a single circuit breaker per distinct connectionstring could be an interesting topic to research at such a point.
Wdyt?

Reply:
I think this wait should be interruptible through cancellation.

Unrelated to that: `Task.Delay` never uses a thread for anything. It starts a timer and completes the `Task` based on that. The `Wait` does use a thread, though.

Reply:
That's a really interesting point, thanks for the suggestion! We're planning to make this aspect completely extensible, and may start introducing a "default behavior" but leaving room for more complex scenarios like the one you mentioned. I'll add this to our list.